PlayerInfo/PlayerInfo.cpp

```diff
@@ -85,19 +89,60 @@ namespace Plugin {
         return message;
     }
 
+    void PlayerInfo::Deactivated(RPC::IRemoteConnection* connection)
+    {
+        if (connection->Id() == _connectionId) {
+            ASSERT(_service != nullptr);
+            // This can potentially be called on a socket thread, so the deactivation (wich in turn kills this object) must be done
+            // on a seperate thread. Also make sure this call-stack can be unwound before we are totally destructed.
+            Core::IWorkerPool::Instance().Submit(PluginHost::IShell::Job::Create(_service, PluginHost::IShell::DEACTIVATED, PluginHost::IShell::FAILURE));
+        }
+    }
+
     /* virtual */ void PlayerInfo::Deinitialize(PluginHost::IShell* service)
     {
         ASSERT(_player != nullptr);
+        ASSERT(service == _service);
+
+        _service->Unregister(&_rcnotification);
+
         if (_player != nullptr) {
+            if (_audioCodecs != nullptr) {
+                _audioCodecs->Release();
+                _audioCodecs = nullptr;
+            }
+            if (_videoCodecs != nullptr) {
+                _videoCodecs->Release();
+                _videoCodecs = nullptr;
+            }
+
             Exchange::JPlayerProperties::Unregister(*this);
-            _player->Release();
+            auto const result = _player->Release();
+
+            if (result != Core::ERROR_DESTRUCTION_SUCCEEDED) {
+                RPC::IRemoteConnection* connection(_service->RemoteConnection(_connectionId));
+
+                // The process can disappear in the meantime...
+                if (connection != nullptr) {
+
+                    // But if it did not dissapear in the meantime, forcefully terminate it. Shoot to kill :-)
+                    connection->Terminate();
+                    connection->Release();
+                }
+            }
         }
         if (_dolbyOut != nullptr)
```

Comment:
I suggest doing this before you do the _player->Release. potentially (depending on the implementation, the dolby implementation might be part of the _player implementation (which is likely since you QueryInterface the _player interface for it) than the result != Core::ERROR_DESTRUCTION_SUCCEEDED will indeed be effecuated due to the fact that _dolbuOutput still holds a reference..
Also, from a programatical point of view you could even code this definsively. Also ASSERT (_videoCodecs != nullptr) and ASSERT (_audioCodecs != nullptr)

Reply:
Right, I removed all checks, and added asserts. Also moved the IDolby unregistration.